modules/platform/nuxeo-platform-mail/src/main/java/org/nuxeo/ecm/platform/mail/action/MessageActionPipeDescriptor.java

```diff
@@ -15,102 +15,105 @@
  *
  * Contributors:
  *     Alexandre Russel
- *
- * $Id$
  */
 
 package org.nuxeo.ecm.platform.mail.action;
 
-import java.util.HashMap;
+import java.util.LinkedHashMap;
 import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
 
+import org.apache.commons.lang3.StringUtils;
 import org.nuxeo.common.xmap.annotation.XNode;
 import org.nuxeo.common.xmap.annotation.XNodeList;
 import org.nuxeo.common.xmap.annotation.XObject;
+import org.nuxeo.common.xmap.registry.XMerge;
+import org.nuxeo.common.xmap.registry.XRegistry;
+import org.nuxeo.common.xmap.registry.XRegistryId;
 
 /**
  * @author Alexandre Russel
  * @author Laurent Doguin
  */
 @XObject("pipe")
+@XRegistry
 public class MessageActionPipeDescriptor {
 
     private static final String START_ACTION = "StartAction";
 
     @XNode("@name")
+    @XRegistryId
     private String name;
 
+    /**
+     * @deprecated since 11.5: should use {@link XMerge#MERGE} instead
+     */
+    @Deprecated(since = "11.5")
     @XNode("@override")
-    private Boolean override;
-
-    private final Map<String, MessageActionDescriptor> actionDescriptorsRegistry = new HashMap<>();
-
-    private MessageActionPipe pipe = new MessageActionPipe();
+    public boolean override;
 
+    // initialized trough setter
     @XNodeList(value = "action", type = MessageActionDescriptor[].class, componentType = MessageActionDescriptor.class)
-    MessageActionDescriptor[] actions;
+    private MessageActionDescriptor[] actions;
 
     public String getName() {
         return name;
     }
 
-    public boolean getOverride() {
-        if (override == null) {
-            return false;
-        }
-        return override;
-    }
-
+    /**
+     * @deprecated since 11.5, only used for internal {@link #getPipe()}.
+     */
+    @Deprecated(since = "11.5")
     public Map<String, MessageActionDescriptor> getActions() {
-        for (MessageActionDescriptor action : actions) {
-            actionDescriptorsRegistry.put(action.getId(), action);
-        }
-        return actionDescriptorsRegistry;
+        return getActionsMap();
     }
 
-    /**
-     * Merge this MessageActionPipeDescriptor with the given one.
-     */
-    public void merge(MessageActionPipeDescriptor descriptor) {
-        for (String actionName : descriptor.getActions().keySet()) {
-            actionDescriptorsRegistry.put(actionName, descriptor.getActions().get(actionName));
-        }
-        pipe = new MessageActionPipe();
+    public Map<String, MessageActionDescriptor> getActionsMap() {
+        return Stream.of(actions)
+                     .collect(Collectors.toMap(MessageActionDescriptor::getId, Function.identity(), (e1, e2) -> e2,
+                             LinkedHashMap::new));
```

Comment:
We have lots of those Map collectors now, maybe a good pattern would be to define static methods in the corresponding descriptors like:

    public static final Collector<...> COLLECTOR_TO_ID_DESC_MAP = Collectors.toMap(MessageActionDescriptor::getId, Function.identity(),
               (e1, e2) -> e2, LinkedHashMap::new);